### DISCUSSION TITLE
Support for Peer to Peer Rollback Netcode

### DISCUSSION THREAD

**Discussion by ValorZard** [o3de/o3de#2022] [Ideas]
There's a lot of great resources on what Rollback Netcode is, specifically [an easy-to-understand text article](https://arstechnica.com/gaming/2019/10/explaining-how-fighting-games-use-delay-based-and-rollback-netcode/) and a great [video](https://youtu.be/0NLe4IpdS1w) or [two](https://www.youtube.com/watch?v=1RI5scXYhK0). (Plus for a MUCH more technical explanation and breakdown, [watch this talk by the people behind Mortal Kombat.](https://youtu.be/7jb0FOcImdg))

But essentially, the idea behind Rollback Netcode is that fighting games require very low input latency. When they play online, players want their inputs to come out as soon as possible, and not have any of it lost to lag. The best way to do this is by compensating for lag. Inputs from one player won't always reach the other player in time, so the game simulates those inputs until those inputs finally reach the other player. When these late inputs reach, the game essentially redos the last 100 milliseconds of the game in a single frame update (16 milliseconds), and the players barely notice it happening, save for a few visual artifacts.

You might think it sounds pretty simple, and it honestly is! However, many fighting game developers based in Japan have not used it yet for a variety of reasons, leading to games becoming unplayable online. American developers have already implemented this however, with a certain open-source library called [GGPO](https://github.com/pond3r/ggpo).

However, I think since o3de already has rollback support in the form of client-server rollback (similar to what [Overwatch](https://www.youtube.com/watch?v=W3aieHjyNvw) or [Halo Reach](https://youtu.be/h47zZrqjgLc) does, it might be simpler and easier to just write up support for peer to peer rollback using what we already have.

Additionally, what is super important is getting determinism down. There's [a great article about it by gaffer on games](https://gafferongames.com/post/deterministic_lockstep/), but rollback netcode is actually a type of deterministic lockstep, where only inputs are being sent. As such, the simulation HAS to be completely deterministic and run the same on every system. 

This isn't as bad as it seems, since there's two way we can do this.

1. Use fixed point
   -  We make a new physics library that only uses fixed points. Fixed points are essentially just ints that are split to have some bits represents decimal places. This will work on all architectures we can think of, but will be very slow.
2. Use soft float
   - A while back, there was this project [that was made in unity using soft floats.](https://github.com/Kimbatt/unity-deterministic-physics) It looks super interesting, and is also deterministic. I am a hundred percent sure you could do something similar in o3de as well using Berkley soft floats.

Even though all of this might sound foreign to those who don't play or make fighting games, I know that the technical know-how to do this is here. Double Helix, who is now a part of Amazon, worked on Killer Instinct Season 1, which has some of the best fighting game netcode in the business. This made sound like a lot and is indeed a specialized use case, but this would make this engine extremely appealing to fighting game developers, and other developers making arcade-style games. This engine is still new, so making a feature like this isn't fully out of scope yet. Are you interested?

**Comment by olegbc1** [comment]
@ValorZard thank you for detailed overview. I am a big fan of fighting games, and especially MK so this is pretty close to home...
I went ahead and created a [feature request ](https://github.com/o3de/o3de/issues/2075)off of your post so @o3de/sig-network can triage.

**Comment by lmbr-pip** [comment]
This seems to have fallen between some gaps. It went from this discussion -> feature request issue which then asked for an RFC and stalled there. Am going to get a few people to look at this and see what the next steps should be. So apologies @ValorZard  for lack of attention here. Aim to have an update soon.
